### PR TITLE
Emacs refactor

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Description of changes

Cosmetical and structural changes to the code.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by Atemu — COMMENTED** [review]
Quick initial screen of the diff.

Generally looks very good, certainly necessary.

File: pkgs/applications/editors/emacs/generic.nix [hunk 1]
```diff
@@ -256,29 +405,41 @@ in
...
+      source code.
     '';
+    license = lib.licenses.gpl3Plus;
+    maintainers = with lib.maintainers; [
+      adisbladis
+      atemu
+      jwiegley
+      lovek323
+      matthewbauer
+    ];
```
[the remark this answers is not in the document]
Comment (AndersonTorres) [reply pkgs/applications/editors/emacs/generic.nix]: Yep!

**Comment by AndersonTorres** [comment]
Is the Darwin X86 machine failing somehow?

**Comment by Atemu** [comment]
They seem to have genuinely taken over an hour and were subsequently timed out. I'll run a local build.

**Comment by Atemu** [comment]
You should prioritise code quality for now though, we can look at build issues later.

**Comment by Atemu** [comment]
FWIW, it builds and runs just fine on aarch64-darwin.

Do you want this to ship in 23.05 or do we want to wait for next release?

**Comment by AndersonTorres** [comment]
Given this is just a structural but not semantic change (except maybe by the removal of `rec`), it can be included in 23.05.

I like it!

**Comment by AndersonTorres** [comment]
Next, I plan to remove the `? null`'s. This will require including Boolean guards like `fancylibSupport ? false, fancylib`.

**Comment by Atemu** [comment]
I think it'd be enough to just remove the null defaults. It's a deprecated pattern and we should first have to evaluate whether we even need the ability to disable those things. Big feature toggles like nativeComp or UI toolkit make sense but I don't think we need the ability to turn of every little library.

**Comment by AndersonTorres** [comment]
This split of sources, `28.nix` and macport.nix`, looks too antinatural to me. It should be better modularized, even because there are at least another interesting fork of Emacs, namely, [Commercial Emacs](https://github.com/commercial-emacs/commercial-emacs), that uses GitHub. Then we have three provenances: bitbucket, savannah and github.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -30157,8 +30155,7 @@ with pkgs;
 
   emacsMacport = callPackage ../applications/editors/emacs/macport.nix {
     withMacport = true;
-
-    gconf = null;
+    gconf = gnome2.GConf;
```
Comment (Atemu) [inline pkgs/top-level/all-packages.nix:30158]: Do we really need gconf on Darwin?

Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix:30158]: This is the way.
gconf is not on the toplevel. Then it needs to receive an explicit value, whether if the underlying code will use it or not.

Comment (Atemu) [reply pkgs/top-level/all-packages.nix]: No, I'm asking whether gconf is in any way required for macport emacs (which only runs on Darwin).

Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix]: I can't know for sure. The Nix code by itself filters gconf when building for Darwin, so my guess is that it isn't needed.

**Comment by AndersonTorres** [comment]
@ofborg build emacsMacport emacs28-nox

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (Atemu) [reply pkgs/top-level/all-packages.nix]: That's what I mean. Just drop the override then.

**Comment by AndersonTorres** [comment]
Now the code looks cute, but the stack of modifications looks ugly.

**Comment by Atemu** [comment]
Just squash in the end. One large refactor commit. That's what I did for ffmpeg.

Keep the history for now and just add commits on top (without force-push) so that we can follow along.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix]: I am thinking about removing gconf entirely, given that it became obsolete/deprecated.

**Comment by AndersonTorres** [comment]
Given that I suggested to include this on 23.05, should this PR be split? Or is it possible to backport only the refactor?

**Comment by adisbladis** [comment]
> Given that I suggested to include this on 23.05, should this PR be split? Or is it possible to backport only the refactor?

I think we can backport the whole thing including the addition of 29. It's perfectly fine to include a pre-release under a separate attribute.
On that topic we should also add an `emacsPgtk` attribute to nixpkgs.

**Comment by AndersonTorres** [comment]
It is preferrable to add `emacs-pgtk` when 29 becomes officially stable.

**Comment by otavio** [comment]
Awesome :-) As an Emacs user, this is indeed a nice rework 👍🏼

**Comment by AndersonTorres** [comment]
@ofborg build emacs-macport  emacs29 emacs29-gtk3 emacs29-nox emacs29-pgtk emacs28 emacs28-gtk2 emacs28-gtk3 emacs28-nox

**Comment by adisbladis** [comment]
I think you accidentally fixed a packaging bug where `gpm` wasn't enabled even though `withGpm` was set to true:
```
- /nix/store/0s3aqz3jz2k59c4sb13l0w9xrwlkr5wm-emacs-28.2.drv:{out}
+ /nix/store/by71r381w7b28zgclpxiapr4rjka31mw-emacs-28.2.drv:{out}
• The set of input derivation names do not match:
    + gpm-unstable-2020-06-17
• The input derivation named `source` differs
  - /nix/store/ymsm3vfq4jjzvn4i4na2v18b9bsjqm70-source.drv:{out}
  + /nix/store/f3ks0dhgixhsdgm8k9sk9ipf5606b3i6-source.drv:{out}
  • The environments do not match:
      - outputHashAlgo=sha256
• Skipping environment comparison
```

It goes to show how bug prone the previous setup was.

**Comment by AndersonTorres** [comment]
It makes me wonder. By reading the old code, gpm was inserted when `stdenv.isLinux` unconditionally. Maybe it was the `= null`...

**Comment by adisbladis** [comment]
It was absolutely the `gpm = null` which made `withGpm` useless without _also_ passing in `gpm` with the override.
Very confusing UX.

**Comment by adisbladis** [comment]
> It is preferrable to add emacs-pgtk when 29 becomes officially stable. After all, afaicr there is no emacs28-pgtk.

That's only because pgtk is a new thing in 29. I'd like for users to have _some_ ability to use pgtk on 23.05, even if it means using a pre-release.
We can always backport the stable release when that finally happens.

We can deal with pgtk separately though.

**Comment by peterhoeg** [comment]
Couple of points:

1. there is pgtk support for emacs 28 although not officially. That being said, I was using it until emacs29-pgtk became stable enough for daily use. We could add this in as an additional option.
2. any reason that sqlite3 and webp support isn't on by default on 29+? (they are upstream)
3. xwidgets is broken with emacs29-pgtk as it depends on `withX` which is false with pgtk so the dependencies are not being added. It otherwise works fine if those dependencies are added.

As for `withX` in general, we need to decide if it means "X11R6 and nothing else" or "something X-like which also includes GTK with wayland".

**Comment by Atemu** [comment]
@adisbladis I think I would have liked a day or two to review the final refactor too.

**Comment by AndersonTorres** [comment]
My intention was to keep the previous semantics, without thinking too much about it. Now the issues are way more visible.

**Comment by AndersonTorres** [comment]
> xwidgets is broken with emacs29-pgtk as it depends on withX which is false with pgtk so the dependencies are not being added. It otherwise works fine if those dependencies are added.

Should PGTK include XWidgets?

> As for withX in general, we need to decide if it means "X11R6 and nothing else" or "something X-like which also includes GTK with wayland". 

I vote for "pure X".

**Comment by AndersonTorres** [comment]
Small bug report: I tried to build pgtk with xwidgets and it complained the lack of webkit.

**Comment by peterhoeg** [comment]
That's what i meant - you need to manually add the buildinputs because
withx is false

**Comment by AndersonTorres** [comment]
Next iteration: https://github.com/NixOS/nixpkgs/pull/235859

**Pull request merged**